services/housetables/src/main/java/com/linkedin/openhouse/housetables/controller/UserHouseTablesController.java

```diff
@@ -189,4 +254,61 @@ public ResponseEntity<Void> renameTable(
     return new ResponseEntity<>(
         apiResponse.getResponseBody(), apiResponse.getHttpHeaders(), apiResponse.getHttpStatus());
   }
+
+  @Operation(
+      summary = "restore a Soft Deleted User Table",
+      description =
+          "Restores an existing soft-deleted User House Table identified by databaseID, tableId, and deletedAtMs. "
+              + " Will fail if the table's databaseId and tableId is currently in use.",
+      tags = {"UserTable"})
+  @ApiResponses(
+      value = {
+        @ApiResponse(responseCode = "204", description = "User Table PUT: NO_CONTENT"),
+        @ApiResponse(responseCode = "400", description = "User Table PUT: BAD_REQUEST"),
+        @ApiResponse(responseCode = "404", description = "User Table PUT: TBL_DB_NOT_FOUND"),
+        @ApiResponse(responseCode = "409", description = "User Table PUT: CONFLICT")
+      })
+  @PutMapping(value = HTS_TABLES_RESTORE_ENDPOINT)
+  public ResponseEntity<EntityResponseBody<UserTable>> restoreUserTable(
+      @RequestParam(value = "databaseId") String databaseId,
+      @RequestParam(value = "tableId") String tableId,
+      @RequestParam(value = "deletedAtMs") Long deletedAtMs) {
+    com.linkedin.openhouse.common.api.spec.ApiResponse<EntityResponseBody<UserTable>> apiResponse;
+    apiResponse =
+        softDeletedTablesHtsApiHandler.restoreEntity(
+            SoftDeletedUserTableKey.builder()
+                .databaseId(databaseId)
+                .tableId(tableId)
+                .deletedAtMs(deletedAtMs)
+                .build());
+    return new ResponseEntity<>(
+        apiResponse.getResponseBody(), apiResponse.getHttpHeaders(), apiResponse.getHttpStatus());
+  }
+
+  @Operation(
+      summary = "Purge a Soft Deleted User Table",
+      description = "Permanently deletes an existing soft-deleted User Table",
+      tags = {"UserTable"})
+  @ApiResponses(
+      value = {
+        @ApiResponse(responseCode = "204", description = "User Table PUT: NO_CONTENT"),
+        @ApiResponse(responseCode = "400", description = "User Table PUT: BAD_REQUEST"),
+        @ApiResponse(responseCode = "404", description = "User Table PUT: TBL_DB_NOT_FOUND")
+      })
+  @DeleteMapping(value = HTS_TABLES_PURGE_ENDPOINT)
+  public ResponseEntity<Void> purgeSoftDeletedUserTable(
+      @RequestParam(value = "databaseId") String databaseId,
+      @RequestParam(value = "tableId") String tableId,
+      @RequestParam(value = "deletedAtMs") Long deletedAtMs) {
```

Comment:
Do we need deletedAtMs here? Purge means should purge all the entries for soft deleted tables for the tableId and databaseId?

Reply:
I was thinking we change this to `purgeFromMs` to delete all table rows with `purgeAfterMs` < `purgeFromMs`, but let `purgeFromMs` be configurable. You're right deleting record by record isn't that useful here. Originally it was to make deletions between HDFS and MySQL more consistent but we discussed performing a mysql purge followed by a bulk HDFS purge of those folders instead, and letting ODD handle any dangling folders if HDFS deletion fails (which would have been needed anyways).